crates/ruff_python_formatter/src/expression/string.rs

```diff
@@ -639,3 +655,301 @@ fn normalize_string(
+        .min()
+        .unwrap_or_default();
+
+    while let Some(line) = lines.next() {
+        let is_last = lines.peek().is_none();
+        format_docstring_line(
+            line,
+            is_last,
+            offset,
+            stripped_indentation,
+            already_normalized,
+            f,
+        )?;
+        // We know that the normalized string has \n line endings
+        offset += line.text_len() + "\n".text_len();
+    }
+
+    // Same special case in the last line as for the first line
+    let trim_end = normalized.as_ref().trim_end();
+    if trim_end.ends_with(string_part.preferred_quotes.style.as_char())
+        || trim_end.chars().rev().take_while(|c| *c == '\\').count() % 2 == 1
+    {
+        space().fmt(f)?;
+    }
+
+    write!(f, [string_part.preferred_quotes])
+}
+
+/// Format a docstring line that is not the first line
+fn format_docstring_line(
+    line: &str,
+    is_last: bool,
+    offset: TextSize,
+    stripped_indentation: TextSize,
+    already_normalized: bool,
+    f: &mut PyFormatter,
+) -> FormatResult<()> {
+    let trim_end = line.trim_end();
+    if trim_end.is_empty() {
+        return if is_last {
+            // If the doc string ends with `    """`, the last line is `    `, but we don't want to
+            // insert an empty line (but close the docstring)
+            Ok(())
+        } else {
+            empty_line().fmt(f)
+        };
+    }
+
+    let tab_or_non_ascii_space = trim_end
+        .chars()
+        .take_while(|c| c.is_whitespace())
+        .any(|c| c != ' ');
+
+    if tab_or_non_ascii_space {
+        // We strip the indentation that is shared with the docstring statement, unless a line
+        // was indented less than the docstring statement, in which we strip only this much
+        // indentation to implicitly pad all lines by the difference, or all lines were
+        // overindented, in which case we strip the additional whitespace (see example in
+        // [`format_docstring`] doc comment). We then prepend the in-docstring indentation to the
+        // string.
+        let indent_len = count_indentation_like_black(trim_end) - stripped_indentation;
+        let in_docstring_indent = " ".repeat(indent_len.to_usize()) + trim_end.trim_start();
+        dynamic_text(&in_docstring_indent, Some(offset)).fmt(f)?;
+    } else {
+        // Take the string with the trailing whitespace removed, then also skip the leading
+        // whitespace
+        let trimmed_line_range =
+            TextRange::at(offset, trim_end.text_len()).add_start(stripped_indentation);
+        if already_normalized {
+            source_text_slice(trimmed_line_range, ContainsNewlines::No).fmt(f)?;
+        } else {
+            // All indents are ascii spaces, so the slicing is correct
+            dynamic_text(
+                &trim_end[stripped_indentation.to_usize()..],
+                Some(trimmed_line_range.start()),
+            )
+            .fmt(f)?;
+        }
```

Comment:
This repeats itself multiple times. I wonder if it would make sense to implement your type that stores the `Cow` internally and has a `write_slice(range, f)` method. It would remove the `is_normalized` boolean. However, I'm not sure if that plays nicely with `lines` (or a `Slice` enum that allows wrapping a &str and `is_normalized).

Reply:
i've though about this but docstring formatting is a special case and i wouldn't overengineer it, it already has enough indirection